### PR TITLE
[Merged by Bors] - feat: `NNRat.cast`

### PR DESCRIPTION
**Pull request by YaelDillies**
Define the canonical coercion from the nonnegative rationals to any division semiring.

From LeanAPAP

---

- [x] depends on: #11202
- [x] depends on: #11453
- [x] depends on: #11480
- [x] depends on: #11485
- [x] depends on: #11486
- [x] depends on: #11499
- [x] depends on: #11503
- [x] depends on: #11504
- [x] depends on: #11507
- [x] depends on: #11508
- [x] depends on: #11639
- [x] depends on: #12360

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Field/Defs.lean [hunk 1]
```diff
@@ -47,93 +45,160 @@ a `GroupWithZero` lemma instead.
 field, division ring, skew field, skew-field, skewfield
 -/
 
+-- `NeZero` should not be needed in the basic algebraic hierarchy.
+-- assert_not_exists NeZero
```
Comment (YaelDillies) [inline Mathlib/Algebra/Field/Defs.lean:49]: @semorrison, how much do you care about this `assert_not_exists`? #11202 (+ previous PRs) is an attempt at keeping it, but that PR not enough and I'm finding it hard to avoid ditching it.

File: Mathlib/Algebra/Field/Opposite.lean [hunk 2]
```diff
@@ -13,64 +13,73 @@ import Mathlib.Data.Int.Cast.Lemmas
...
 
 @[to_additive (attr := simp, norm_cast)]
-theorem unop_ratCast [RatCast α] (q : ℚ) : unop (q : αᵐᵒᵖ) = q :=
-  rfl
+lemma unop_ratCast [RatCast α] (q : ℚ) : unop (q : αᵐᵒᵖ) = q := rfl
 #align mul_opposite.unop_rat_cast MulOpposite.unop_ratCast
 #align add_opposite.unop_rat_cast AddOpposite.unop_ratCast
 
-variable (α)
+instance instDivisionSemiring [DivisionSemiring α] : DivisionSemiring αᵐᵒᵖ where
```
Comment (eric-wieser) [inline Mathlib/Algebra/Field/Opposite.lean:40]: This instance renaming (and any others in this PR) would be better as a split

File: Mathlib/Algebra/Field/Defs.lean (hunk 1 above)
Comment (YaelDillies) [reply Mathlib/Algebra/Field/Defs.lean]: Nevermind, I managed to fix it

File: Mathlib/Algebra/Field/Defs.lean [hunk 3]
```diff
@@ -47,93 +45,180 @@ a `GroupWithZero` lemma instead.
...
+class DivisionSemiring (α : Type*) extends Semiring α, GroupWithZero α, NNRatCast α where
+  protected nnratCast := NNRat.castRec
+  /-- However `NNRat.cast` is defined, it must be equal to `a / b`.
+
+  Do not use this lemma directly. Use `NNRat.cast_def` instead. -/
+  protected nnratCast_def (q : ℚ≥0) : (NNRat.cast q : α) = q.num / q.den := by intros; rfl
+  /-- Scalar multiplication by a nonnegative rational number.
+
+  Do not use directly. Instead use the `•` notation. -/
+  protected nnqsmul : ℚ≥0 → α → α := nnqsmulRec NNRat.cast
```
Comment (eric-wieser) [inline Mathlib/Algebra/Field/Defs.lean:117]: If #11262 goes ahead, this should change to
```suggestion
  protected nnqsmul : ℚ≥0 → α → α
```
along with a comment advising to use `nnqsmulRec _`.

File: Mathlib/Algebra/Field/Opposite.lean (hunk 2 above)
Comment (eric-wieser) [reply Mathlib/Algebra/Field/Opposite.lean:41]: #11453 is part of this

File: Mathlib/Algebra/Ring/Int.lean
Comment (eric-wieser) [inline Mathlib/Algebra/Ring/Int.lean:1]: This looks like an easy spin-off PR

File: Mathlib/Tactic/NormNum/Basic.lean [hunk 4]
```diff
@@ -101,7 +101,7 @@ theorem isNat_natAbs_neg : {n : ℤ} → {a : ℕ} → IsInt n (.negOfNat a) →
 
 /-! # Casts -/
 
-theorem isnatCast {R} [AddMonoidWithOne R] (n m : ℕ) :
+theorem isNat_natCast {R} [AddMonoidWithOne R] (n m : ℕ) :
```
Comment (eric-wieser) [inline Mathlib/Tactic/NormNum/Basic.lean:104]: Probably also doesn't belong in this PR

File: Mathlib/Data/NNRat/Defs.lean [hunk 5]
```diff
@@ -25,6 +25,11 @@ of `x` with `↑x`. This tactic also works for a function `f : α → ℚ` with
 ## Notation
 
 `ℚ≥0` is notation for `NNRat` in locale `NNRat`.
+
+## Huge warning
+
+Whenever you state a lemma about the coercion `ℚ≥0 → ℚ`, check that Lean inserts `NNRat.cast`, not
+`Subtype.val`. Else your lemma will never apply.
```
Comment (eric-wieser) [inline Mathlib/Data/NNRat/Defs.lean:32]: Ouch, this is rather unfortunate.

Comment (YaelDillies) [reply Mathlib/Data/NNRat/Defs.lean:32]: Yeah... The precise issue is that `⟨q, hq⟩` notation creates a term of type `{q : ℚ // 0 ≤ q}`, not `ℚ≥0`, so it finds `Subtype.val` as the coercion, not `NNRat.cast`

File: Mathlib/Algebra/Ring/Int.lean
Comment (YaelDillies) [reply Mathlib/Algebra/Ring/Int.lean:1]: #12235

File: Mathlib/Tactic/NormNum/Basic.lean (hunk 4 above)
Comment (YaelDillies) [reply Mathlib/Tactic/NormNum/Basic.lean]: #12236

**Comment by sgouezel** [comment]
!bench

**Comment by sgouezel** [comment]
The bench is very good: I was afraid of a global slowdown, but there is nothing in sight. I don't think the slowdown in two specific files has any importance (although if you can understand what is going on this would be even better).

**Comment by YaelDillies** [comment]
Yes, I already removed the problematic instances like `Module ℚ α → Module ℚ≥0 α`.

For `FieldTheory.Subfield`, I just checked and it's indeed simply due to the `DivisionRing` instance having more data. For `LinearAlgebra.Matrix.SpecialLinearGroup`, nothing is sticking out as suspicious to me.

File: Mathlib/Data/Rat/Cast/Defs.lean [hunk 6]
```diff
@@ -207,6 +289,19 @@ instance Rat.subsingleton_ringHom {R : Type*} [Semiring R] : Subsingleton (ℚ 
 
 /-! ### Scalar multiplication -/
 
+namespace NNRat
+variable [DivisionSemiring α]
+
+instance (priority := 100) instDistribSMul : DistribSMul ℚ≥0 α where
+  smul := (· • ·)
```
Comment (eric-wieser) [inline Mathlib/Data/Rat/Cast/Defs.lean:296]: I think it's best to leave out this field?

Comment (YaelDillies) [reply Mathlib/Data/Rat/Cast/Defs.lean:296]: Well spotted

**Review by Ruben-VandeVelde — APPROVED** [review]
Code looks fine, assuming this is a change we want.

maintainer merge

File: Mathlib/Data/NNRat/Defs.lean (hunk 5 above)
Comment (fpvandoorn) [reply Mathlib/Data/NNRat/Defs.lean:32]: This suggests we might want an `NNRat.mk` that has the correct return type?

But if so, that can wait until a later PR

bors merge

Comment (YaelDillies) [reply Mathlib/Data/NNRat/Defs.lean:32]: Maybe, but it won't have the constructor notation. To properly solve it we should either:
* make `NNRat` reducible
* make `NNRat` a one-field structure
* make `NNRat` a fully fledged structure with `num den : Nat` etc
